File: src/data/catalog.ts
import { z } from "zod";
import rawCatalog from "../../config/data-catalog.json";

const sourceSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  kind: z.string().min(1),
  url: z.string().min(1),
  provides: z.array(z.string().min(1)).min(1),
  license: z.string().min(1),
  usedBy: z.array(z.string().min(1)).min(1),
});

const artifactSchema = z.object({
  id: z.string().min(1),
  label: z.string().min(1),
  path: z.string().min(1),
  version: z.string().min(1),
});

const jobSchema = z.object({
  id: z.string().regex(/^[a-z0-9-]+$/),
  title: z.string().min(1),
  schedule: z.string().min(5),
  frequency: z.string().min(1),
  deadlineSeconds: z.number().int().positive(),
  command: z.string().min(1),
  args: z.array(z.string()),
  sourceIds: z.array(z.string()).min(1),
  artifactIds: z.array(z.string()).min(1),
  profile: z.enum(["standard", "landscape", "inference"]),
});

const catalogSchema = z.object({
  schemaVersion: z.literal(1),
  catalogVersion: z.string().min(1),
  timeZone: z.string().min(1),
  runtime: z.object({
    pipelineVersion: z.string().min(1),
    profilePipelineVersion: z.string().min(1),
    scenePromptVersion: z.string().min(1),
    sceneReconcilerVersion: z.string().min(1),
    osmPbfUrl: z.url(),
    geoAdminBaseUrl: z.url(),
    geoAdminDataBaseUrl: z.url(),
    transportApiBaseUrl: z.url(),
    graphHopperDefaultUrl: z.url(),
    mapStyleUrl: z.url(),
    mapRasterTileUrl: z.string().url().includes("{z}"),
    landCoverVersion: z.string().min(1),
    landCoverTileUrl: z.url(),
  }),
  providers: z.object({
    gtfsCatalogueBaseUrl: z.url(),
    gtfsDownloadHosts: z.array(z.string().min(1)).min(1),
    geoAdminHeightUrl: z.url(),
    geoAdminProfileUrl: z.url(),
    swissTlmItemsUrl: z.url(),
    swissBuildingsItemsUrl: z.url(),
    swisstopoRasterItemsTemplate: z.string().url().includes("{collection}"),
    swissAltiCollection: z.string().min(1),
    swissSurfaceCollection: z.string().min(1),
    meteoIconCollection: z.string().min(1),
    meteoIconStacCollection: z.string().min(1),
    meteoIconHorizontalConstants: z.string().min(1),
    meteoRadarItemsUrl: z.url(),
    meteoStationMetadataUrl: z.url(),
    meteoStationCurrentTemplate: z.string().url().includes("{station}"),
    panoramaxSearchUrl: z.url(),
    panoramaxViewerUrl: z.url(),
    commonsApiUrl: z.url(),
    kartaViewNearbyUrl: z.url(),
    kartaViewViewerUrl: z.url(),
    swissImageWmsUrl: z.url(),
    swissImageMapUrl: z.url(),
    swissImageLayer: z.string().min(1),
    inferenceDefaultUrl: z.url(),
  }),
  sources: z.array(sourceSchema).min(1),
  artifacts: z.array(artifactSchema).min(1),
  jobs: z.array(jobSchema).min(1),
}).superRefine((catalog, context) => {
  const checkUnique = (values: string[], label: string) => {
    if (new Set(values).size !== values.length) context.addIssue({ code: "custom", message: `${label} müssen eindeutig sein` });
  };
  checkUnique(catalog.sources.map(({ id }) => id), "Quellen-IDs");
  checkUnique(catalog.artifacts.map(({ id }) => id), "Artefakt-IDs");
  checkUnique(catalog.jobs.map(({ id }) => id), "Job-IDs");

  const sources = new Set(catalog.sources.map(({ id }) => id));
  const artifacts = new Set(catalog.artifacts.map(({ id }) => id));
  for (const job of catalog.jobs) {
    for (const id of job.sourceIds) if (!sources.has(id)) context.addIssue({ code: "custom", message: `Job ${job.id} referenziert unbekannte Quelle ${id}` });
    for (const id of job.artifactIds) if (!artifacts.has(id)) context.addIssue({ code: "custom", message: `Job ${job.id} referenziert unbekanntes Artefakt ${id}` });
  }
});

export type DataCatalog = z.infer<typeof catalogSchema>;
export type DataSource = DataCatalog["sources"][number];
export type DataJob = DataCatalog["jobs"][number];

export const dataCatalog: DataCatalog = catalogSchema.parse(rawCatalog);

export function sourcesFor(job: DataJob): DataSource[] {
  const ids = new Set(job.sourceIds);
  return dataCatalog.sources.filter(({ id }) => ids.has(id));
}

File: worker/benchly/catalog.py
"""Validated access to Benchly's single data-source and refresh catalog."""

from __future__ import annotations

import json
import os
from functools import lru_cache
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class RuntimeConfig(StrictModel):
    pipelineVersion: str
    profilePipelineVersion: str
    scenePromptVersion: str
    sceneReconcilerVersion: str
    osmPbfUrl: HttpUrl
    geoAdminBaseUrl: HttpUrl
    geoAdminDataBaseUrl: HttpUrl
    transportApiBaseUrl: HttpUrl
    graphHopperDefaultUrl: HttpUrl
    mapStyleUrl: HttpUrl
    mapRasterTileUrl: str
    landCoverVersion: str
    landCoverTileUrl: HttpUrl


class ProviderConfig(StrictModel):
    gtfsCatalogueBaseUrl: HttpUrl
    gtfsDownloadHosts: list[str] = Field(min_length=1)
    geoAdminHeightUrl: HttpUrl
    geoAdminProfileUrl: HttpUrl
    swissTlmItemsUrl: HttpUrl
    swissBuildingsItemsUrl: HttpUrl
    swisstopoRasterItemsTemplate: str
    swissAltiCollection: str
    swissSurfaceCollection: str
    meteoIconCollection: str
    meteoIconStacCollection: str
    meteoIconHorizontalConstants: str
    meteoRadarItemsUrl: HttpUrl
    meteoStationMetadataUrl: HttpUrl
    meteoStationCurrentTemplate: str
    panoramaxSearchUrl: HttpUrl
    panoramaxViewerUrl: HttpUrl
    commonsApiUrl: HttpUrl
    kartaViewNearbyUrl: HttpUrl
    kartaViewViewerUrl: HttpUrl
    swissImageWmsUrl: HttpUrl
    swissImageMapUrl: HttpUrl
    swissImageLayer: str
    inferenceDefaultUrl: HttpUrl

    @model_validator(mode="after")
    def validate_templates(self) -> "ProviderConfig":
        if "{collection}" not in self.swisstopoRasterItemsTemplate:
            raise ValueError("swisstopoRasterItemsTemplate must contain {collection}")
        if "{station}" not in self.meteoStationCurrentTemplate:
            raise ValueError("meteoStationCurrentTemplate must contain {station}")
        return self


class DataSource(StrictModel):
    id: str = Field(pattern=r"^[a-z0-9-]+$")
    name: str
    kind: str
    url: str
    provides: list[str] = Field(min_length=1)
    license: str
    usedBy: list[str] = Field(min_length=1)


class Artifact(StrictModel):
    id: str = Field(pattern=r"^[a-z0-9-]+$")
    label: str
    path: str
    version: str


class DataJob(StrictModel):
    id: str = Field(pattern=r"^[a-z0-9-]+$")
    title: str
    schedule: str
    frequency: str
    deadlineSeconds: int = Field(gt=0)
    command: str
    args: list[str]
    sourceIds: list[str] = Field(min_length=1)
    artifactIds: list[str] = Field(min_length=1)
    profile: str = Field(pattern=r"^(standard|landscape|inference)$")


class DataCatalog(StrictModel):
    schemaVersion: int = Field(ge=1)
    catalogVersion: str
    timeZone: str
    runtime: RuntimeConfig
    providers: ProviderConfig
    sources: list[DataSource] = Field(min_length=1)
    artifacts: list[Artifact] = Field(min_length=1)
    jobs: list[DataJob] = Field(min_length=1)

    @model_validator(mode="after")
    def validate_graph(self) -> "DataCatalog":
        source_ids = [source.id for source in self.sources]
        artifact_ids = [artifact.id for artifact in self.artifacts]
        job_ids = [job.id for job in self.jobs]
        if len(source_ids) != len(set(source_ids)) or len(artifact_ids) != len(set(artifact_ids)) or len(job_ids) != len(set(job_ids)):
            raise ValueError("Catalog IDs must be unique")
        known_sources = set(source_ids)
        known_artifacts = set(artifact_ids)
        for job in self.jobs:
            unknown_sources = set(job.sourceIds) - known_sources
            unknown_artifacts = set(job.artifactIds) - known_artifacts
            if unknown_sources or unknown_artifacts:
                raise ValueError(f"Job {job.id} has unknown references: {unknown_sources | unknown_artifacts}")
            if len(job.schedule.split()) != 5:
                raise ValueError(f"Job {job.id} has an invalid five-field cron expression")
        return self


def catalog_path() -> Path:
    configured = os.environ.get("BENCHLY_DATA_CATALOG")
    return Path(configured) if configured else Path(__file__).resolve().parents[2] / "config" / "data-catalog.json"


@lru_cache(maxsize=1)
def load_catalog() -> DataCatalog:
    with catalog_path().open(encoding="utf-8") as handle:
        return DataCatalog.model_validate(json.load(handle))
